fe/src/main/java/org/apache/doris/common/proc/CurrentQueryInfoProvider.java

```diff
@@ -0,0 +1,518 @@
+        final Map<TNetworkAddress, Request> requestMap = Maps.newHashMap();
+        final Map<TNetworkAddress, TNetworkAddress> brpcAddressMap = Maps.newHashMap();
+        for (QueryStatisticsItem item : items) {
+            for (QueryStatisticsItem.FragmentInstanceInfo instanceInfo : item.getFragmentInstanceInfos()) {
+                // use brpc address
+                TNetworkAddress brpcNetAddress = brpcAddressMap.get(instanceInfo.getAddress());
+                if (brpcNetAddress == null) {
+                    try {
+                        brpcNetAddress = toBrpcHost(instanceInfo.getAddress());
+                        brpcAddressMap.put(instanceInfo.getAddress(), brpcNetAddress);
+                    } catch (Exception e) {
+                        LOG.warn(e.getMessage());
+                        throw new AnalysisException(e.getMessage());
+                    }
+                }
+                // merge different requests
+                Request request = requestMap.get(brpcNetAddress);
+                if (request == null) {
+                    request = new Request(brpcNetAddress);
+                    requestMap.put(brpcNetAddress, request);
+                }
+                // specified query instance which will report.
+                if (!allQuery) {
+                    final PUniqueId pUId = new PUniqueId(instanceInfo.getInstanceId());
+                    request.addInstanceId(pUId);
+                }
+            }
+        }
+        recvResponse(sendRequest(requestMap));
+    }
+
+    private List<Pair<Request, Future<PTiggerReportProfileResult>>> sendRequest(
+            Map<TNetworkAddress, Request> requestMap) throws AnalysisException {
+        final List<Pair<Request, Future<PTiggerReportProfileResult>>> futures = Lists.newArrayList();
+        for (TNetworkAddress address : requestMap.keySet()) {
+            final Request request = requestMap.get(address);
+            final PTiggerReportProfileRequest pbRequest =
+                    new PTiggerReportProfileRequest(request.getInstanceIds());
+            try {
+                futures.add(Pair.create(request, BackendServiceProxy.getInstance().
+                        triggerReportProfileAsync(address, pbRequest)));
+            } catch (RpcException e) {
+                throw new AnalysisException("Sending request fails for query's execution informations.");
+            }
+        }
+        return futures;
+    }
+
+    private void recvResponse(List<Pair<Request, Future<PTiggerReportProfileResult>>> futures)
+            throws AnalysisException {
+        final String reasonPrefix = "Fail to receive result.";
+        for (Pair<Request, Future<PTiggerReportProfileResult>> pair : futures) {
+            try {
+                final PTiggerReportProfileResult result
+                        = pair.second.get(10, TimeUnit.SECONDS);
```

Comment:
You wait here for at most 10 seconds. But getWaitTime() is at most 2 seconds? What's that mean?

Reply:
I will fix it.

Reply:
the for_each of map return Pair class value as function result, so it can't be refered as left-value reference.